packages/fxa-auth-server/lib/oauth/jwt_id_token.js

```diff
@@ -0,0 +1,115 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const AppError = require('./error');
+const jwt = require('./jwt');
+
+const HEADER_TYP = 'at+JWT';
+
+/**
+ * Verify an OIDC ID Token, following the flow in 3.1.3.7 of the spec:
+ * https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation
+ *
+ * @param {String} idToken
+ * @param {String} clientId
+ * @throws `invalidToken` error if ID Token is invalid.
+ * @returns {Promise<Object>} resolves with claims in Token
+ */
+exports.verify = async function verify(idToken, clientId) {
+  // Steps 1 and 2 of the verification flow are already handled by jwt.js:
+  //
+  // 1. if the ID Token is encrypted, decrypt it using the keys and algorithms
+  //    negotiated at Registration time between OpenID Provider and Client.
+  //    If encryption was negotiated but the token is not encrypted, RP SHOULD
+  //    reject it.
+  // 2. The issuer identifier for the OpenID Provider MUST exactly match the
+  //    'iss' Claim.
+  let claims;
+  try {
+    claims = await jwt.verify(idToken, { typ: HEADER_TYP });
+  } catch (err) {
+    throw AppError.invalidToken();
+  }
+
+  // TODO: what is the canonical list of trusted audiences? Do we ever issue
+  // ID Tokens with multiple audiences?
+  //
+  // 3. Client MUST validate that the 'aud' Claim contains its 'client_id'
+  // value registered at the Issuer identifier by the 'iss' Claim as an
+  // audience. If the ID Token doesn't list the Client as a valid audience,
+  // or if the ID Token contains additional audiences not trusted by the
+  // Client, the ID Token MUST be rejected.
+  if (typeof claims.aud === 'string' && claims.aud !== clientId) {
+    throw AppError.invalidToken();
+  } else if (!claims.aud.includes(clientId)) {
+    throw AppError.invalidToken();
+  }
+
+  // Skipping these steps, because we never use the 'azp' claim:
+  //
+  // 4. If the ID Token contains multiple audiences, the Client SHOULD verify
+  //    that an 'azp' field is present.
+  // 5. If an 'azp' Claim is present, the Client SHOULD verify that its
+  //    client_id is the Claim Value.
+
+  // Skipping this optional step:
+  //
+  // 6. Optionally, TLS server validation can be used if the token was received
+  // directly by the Client from the Token Endpoint.
+
+  // TODO: is this always RS256? AFAICT yes...
+  //
+  // 7. The 'alg' value should be the default of 'RS256' or the value of
+  //    'id_token_signed_response_alg' set at Registration.
+  if (claims.alg !== 'RS256') {
+    throw AppError.invalidToken();
+  }
```

Comment:
FWIW, `jwt.verify` will have performed this check already, but I understand why you want to do it here for completeness. You could perhaps make it more explicit by passing in an `algorithms` option to the call to `jwt.verify` above.